src/main/java/org/candlepin/model/OwnerProductCurator.java

```diff
@@ -853,6 +855,10 @@ public void updateOwnerProductReferences(Owner owner, Map<String, String> produc
             return;
         }
 
+        // Flush because any newly created products MUST be sent to the DB before the next calls
+        // as newly created products
+        this.getEntityManager().flush();
```

Comment:
Couple things here:
First, we can call `this.flush()` instead of fetching the entity manager ourselves.

When it was first brought up that we needed to flush, I assumed that was meaning *after* the operations in this method. If the flush needs to occur *before*, that's still fine, but it means we can actually drop a flush in the `ProductNodeVisitor` (and `ContentNodeVisitor` right before this is called.

Reply:
Hmm... it's fair, I like the flush being handled outside of this as the visitors can call flush once before iterating over the maps. One concern with keeping the flush here would be that it is going to be called many times in a row by the ProductNodeVisitor.complete() . It's not horrible as a bunch of the flush calls should no-op but it's an odd contention in how the method is used and where we want to be calling flush from. Seems like either we should remove the flush from product node visitor and keep it here. Or, remove the flush from here and add it to ProductManager.updateProduct() so that both places that call updateOwnerProductReferences() have flushed beforehand. Thoughts?

Reply:
I'm leaning toward adding it to the places in the managers where we call it directly, but we would still want to throw that note in the documentation on the `updateOwner-entity-References` so it's not a gotcha months from now.

As a note, `ProductNodeVisitor.complete` is called once per refresh, so we shouldn't get a ton of repeat flushes no matter where they end up. The confusing bit might be the loop over the map, but that should be a single-entry map in every case I can think of.